File: src/main/java/com/venus/meetspace/common/result/Result.java
package com.venus.meetspace.common.result;

import com.venus.meetspace.common.enums.ResultCode;
import lombok.Data;

@Data
public class Result<T> {

    private int code;

    private String msg;

    private T data;

    public static <T> Result<T> success(T data) {
        Result<T> result = new Result<>();
        result.code = ResultCode.SUCCESS.getValue();
        result.data = data;
        return result;
    }

    public static <T> Result<T> success(T data, String msg) {
        Result<T> result = new Result<>();
        result.code = ResultCode.SUCCESS.getValue();
        result.msg = msg;
        result.data = data;
        return result;
    }

    public static <T> Result<T> fail(String msg, ResultCode code) {
        Result<T> result = new Result<>();
        result.code = code.getValue();
        result.msg = msg;
        return result;
    }
}

File: frontend/lib/model/result.dart
import 'package:json_annotation/json_annotation.dart';

part 'result.g.dart';

enum ResultCode { SUCCESS, FAIL, UNAUTHORIZED, FORBIDDEN, NOT_FOUND, DUPLICATION, VALUE_ERROR, NO_SUCH_OBJECT, STATUS_ERROR, INTERNAL_SERVER_ERROR }

@JsonSerializable(genericArgumentFactories: true)
class Result<T> {
  final ResultCode code;
  final String? msg;
  final T? data;

  Result({required this.code, this.msg, this.data});

  factory Result.success(T data, {String? msg}) => Result(code: ResultCode.SUCCESS, msg: msg, data: data);
  factory Result.fail(T data, {String? msg}) => Result(code: ResultCode.FAIL, msg: msg, data: data);
  factory Result.error(ResultCode code, T data, {String? msg}) => Result(code: code, msg: msg, data: data);

  factory Result.fromJson(
      Map<String, dynamic> json, T Function(Object? json) fromJsonT) =>
      _$ResultFromJson(json, fromJsonT);

  Map<String, dynamic> toJson(Object? Function(T value) toJsonT) =>
      _$ResultToJson(this, toJsonT);

  bool get isSuccess => code == ResultCode.SUCCESS;
  
  @override
  String toString() => 'Result{code: $code, msg: $msg, data: $data}';
}
